### PR TITLE
feat: multihash-implementation

### PR DESCRIPTION
**Pull request by Ben-Cheng**
Fixes: #70 

### Change Description:

...

### Notes for Testing:

...

#### Source Code Documentation Tasks:
- [ ] README updated (where applicable)
- [ ] CHANGELOG (Unreleased section) updated
- [ ] Docstrings / comments included to help explain code

#### User Documentation Tasks:
- [ ] Confluence updated (where applicable)

#### Testing Tasks:
- [ ] Added tests that fail without this change
- [ ] All tests are passing in development environment
- [ ] Reviewers assigned
- [ ] Linked to main issue for ZenHub board

### CODE REVIEW

File: topo_processor/util/checksum.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import multihash
+
+
+def multihash_as_hex(file_path: str) -> str:
```
Comment (palmerj) [inline topo_processor/util/checksum.py:4]: Might be good to read the file in chunks. See https://github.com/linz/linz-data-lake/blob/master/data_lake/core/format/gdal_tile.py#L192

Comment (Ben-Cheng) [reply topo_processor/util/checksum.py:4]: what are the benefits of doing that?

Comment (palmerj) [reply topo_processor/util/checksum.py:4]: Will reduce amount of memory required. Good practise when dealing with largish files.

File: topo_processor/util/checksum.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import hashlib
+
+import multihash
+
+
+def multihash_as_hex(file_path, hash_method: str = "sha2-256") -> str:
```
Comment (blacha) [inline topo_processor/util/checksum.py:6]: `file_path` missing type
`hash_method` is not respected as you  just `hashlib.sha256()`

`BLOCK_SIZE` should not be defined inside the function


The actual iteratorion of the file could be simplified to
```
with open('file.name', 'rb') as file:
    while chunk := file.read(CHUNK_SIZE)
        file_hash.update(chunk)
```

File: topo_processor/metadata/loaders/metadata_loader_tiff.py [hunk 3]
```diff
@@ -16,6 +17,7 @@ def is_applicable(self, item: Item) -> bool:
 
     def add_metadata(self, item: Item) -> None:
         # TODO use GDAL to load metadata from the tiff file if it exists
+        item.stac_item.properties["checksum:multihash"] = multihash_as_hex(item.path)
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_tiff.py:20]: this should be in its own loader

`medata_loader_checksum.py`

**Comment by blacha** [comment]
also PR title needs to be conventional commits `Feat/multihash-implementation` is not conventional

File: topo_processor/util/checksum.py (hunk 2 above)
Comment (Ben-Cheng) [reply topo_processor/util/checksum.py:6]: wow, good to know what `:=` does

Comment (palmerj) [reply topo_processor/util/checksum.py:6]: := is new in python 3.8 - and a great improvement it is :-)

File: topo_processor/metadata/loaders/metadata_loader_checksum.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class MetadataLoaderChecksum(MetadataLoader):
+    name = "loader.checksum"
+
+    def is_applicable(self, item: Item) -> bool:
+        return is_tiff(item.path)
+
+    def add_metadata(self, item: Item) -> None:
+        # TODO use GDAL to load metadata from the tiff file if it exists
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_checksum.py:15]: This TODO is not related to this?

File: topo_processor/metadata/item.py [hunk 5]
```diff
@@ -19,6 +19,7 @@ def __init__(self, path: str, collection: Collection):
         self.collection = collection
         self.stac_item = stac.Item(
             # TODO edit these item fields, add collection back to the constuctor:stac_item after collection.stac_collection is implemented
+            stac_extensions=["checksum"],
```
Comment (blacha) [inline topo_processor/metadata/item.py:22]: This should be added in the checksum loader section

File: topo_processor/metadata/loaders/metadata_loader_checksum.py (hunk 4 above)
Comment (Ben-Cheng) [reply topo_processor/metadata/loaders/metadata_loader_checksum.py:15]: no, I will remove it

File: topo_processor/metadata/item.py (hunk 5 above)
Comment (Ben-Cheng) [reply topo_processor/metadata/item.py:22]: good point

File: topo_processor/metadata/loaders/metadata_loader_checksum.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class MetadataLoaderChecksum(MetadataLoader):
+    name = "loader.checksum"
+
+    def is_applicable(self, item: Item) -> bool:
+        return is_tiff(item.path)
+
+    def add_metadata(self, item: Item) -> None:
+        item.stac_item.stac_extensions.append("checksum")
```
Comment (blacha) [inline topo_processor/metadata/loaders/metadata_loader_checksum.py:15]: you need to check if it already exists.

Comment (Ben-Cheng) [reply topo_processor/metadata/loaders/metadata_loader_checksum.py:15]: but isn't this the only place we are appending checksum?

**Review by blacha — APPROVED** [review]
